package galaxy.ins;

option cc_generic_services = true;

message Entry {
    required string key = 1;
    required string value = 2;
}

message AppendEntriesRequest {
    required int64 term = 1;
    required string leader_id = 2;
    optional int64 prev_log_index = 3;
    optional int64 prev_log_term = 4;
    optional int64 leader_commit_index = 5;
    repeated Entry entries = 6;
}

message AppendEntriesResponse {
    required int64 current_term = 1;
    required bool success = 2;
}

message VoteRequest {
    required int64 term = 1;
    required string candidate_id = 2;
    optional int64 last_log_index = 3;
    optional int64 last_log_term = 4;
}

message VoteResponse {
    required int64 term = 1;
    required bool vote_granted = 2;
}

message PutRequest {
    required string key = 1;
    required string value = 2;    
}

message PutResponse {
    required bool success = 1;
}

message GetRequest {
    required string key = 1;    
}

message GetResponse {
    required bool hit = 1;
    required string value = 2;
}

service InsNode {
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    rpc Vote(VoteRequest) returns (VoteResponse);
    rpc Put(PutRequest) returns (PutResponse);
    rpc Get(GetRequest) returns (GetResponse);
}

